price-estimator/src/models/markets_results.rs

```diff
@@ -25,6 +26,45 @@ impl From<&pricegraph::TransitiveOrderbook> for MarketsResult {
     }
 }
 
+impl MarketsResult {
+    pub fn into_base_units(
+        self,
+        base_token_info: &TokenBaseInfo,
+        quote_token_info: &TokenBaseInfo,
+    ) -> Self {
+        return Self {
+            asks: into_transitive_orders_in_base_unit(
+                self.asks,
+                &base_token_info,
+                &quote_token_info,
+            ),
+            bids: into_transitive_orders_in_base_unit(
+                self.bids,
+                &base_token_info,
+                &quote_token_info,
+            ),
+        };
+    }
+}
+
+fn into_transitive_orders_in_base_unit(
+    orders: Vec<TransitiveOrder>,
```

Comment:
Instead of taking Vec<TransitiveOrder> and returning Vec<TransitiveOrder> this function could take &mut [TransitiveOrder] but if you prefer this it's also ok.

Reply:
I agree, I would prefer in-place mutation here with `iter_mut()`.

Reply:
From other programming languages I tend to prefer methods without side effects, but I guess since rust makes the mutation very explicit it's really the same here (or is there any benefit in not consuming it). Will change.

Reply:
> From other programming languages I tend to prefer methods without side effects

In those cases, I would slightly prefer passing in an `impl IntoIterator<Item = TransitiveOrder>` and return an `impl Iterator<Item = TransitiveOrder>`. This way the caller doesn't necessarily need to a) privide a `Vec`, but any iterator over transitive orders and b) the caller isn't forced to collect a `Vec`, but can theoretically collect into a `HashMap` of price ranges (a contrived example). However, if you are forcing the caller to give ownership of a `Vec` to return a new one, then I would argue the extra allocation is unnecessary.

One long term advantage of the `fn (IntoIterator) -> Iterator` way is that we may be able to implement transitive order grouping as an iterator, then we wouldn't need any intermediate allocations, but instead compose these iterators and collect once at the end :smile:

Reply:
It's definitely a preference which is why I said i'm fine with either way. Whether the function takes one mutable argument or returns one value is mostly equivalent but by taking a mut slice reference it is clear that the number of elements doesn't change.
On the other hand I'm not sure I like introducing generics until they are needed. It's good to keep it in mind that the function can be generict that way but if we know it's only called in one place and that place is a vector then I usually prefer it to not be generic yet.

Reply:
> Whether the function takes one mutable argument or returns one value is mostly equivalent

But... my poor memory allocator!

> It's good to keep it in mind that the function can be generict that way but if we know it's only called in one place and that place is a vector then I usually prefer it to not be generic yet.

This is a very good point! To be clear I was just trying to illustrate an example where the "pure functional" way to write this function could be beneficial and could produce equally optimized code as with the mutable reference and be more composable than with mutable references or fixed types.

For this case, I think the `&mut [_]` is the way to go.